src/Flexdis86/Disassembler.hs

```diff
@@ -600,17 +742,285 @@ getRMTable modRM mtbl =
-    SkipModRM pfx df ->
-        finish <$> traverse (parseValueType pfx (prefixOperandSizeConstraint pfx df) Nothing) (df^.defOperands)
+disassembleInstruction tr0 = loopPrefixBytes Seq.empty
+  where
+    -- Parse as many bytes as possible that are known to be prefix bytes. Once
+    -- we encounter the first byte that isn't a prefix byte, move on to parsing
+    -- opcode bytes.
+    loopPrefixBytes :: Seq.Seq Word8 -> m InstructionInstance
+    loopPrefixBytes prefixBytes = do
+      b <- readByte
+      -- Although we use Data.List.lookup below, this is acceptable since the
+      -- lists being looked up are very small in length.
+      if |  b `elem` simplePrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  b `elem` segPrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  b `elem` rexPrefixBytes
+         -> loopPrefixBytes (prefixBytes Seq.|> b)
+
+         |  otherwise
+         -> loopOpcodes (F.toList prefixBytes) tr0 b
+
+    -- Parse opcode byte and use them to navigate through the opcode table.
+    -- Once we have parsed all of the opcode bytes in an instruction,
+    -- disassemble the rest of the instruction.
+    loopOpcodes :: [Word8]
+                -> NextOpcodeTable
+                -> Word8
+                -> m InstructionInstance
+    loopOpcodes prefixBytes = go
+      where
+        go :: NextOpcodeTable -> Word8 -> m InstructionInstance
+        go tr opcodeByte =
+          case tr V.! fromIntegral opcodeByte of
+            OpcodeTable tr' -> do
+              opcodeByte' <- readByte
+              go tr' opcodeByte'
+            OpcodeTableEntry defsWithModRM defsWithoutModRM
+              |  -- Check the instruction candidates without a ModR/M byte
+                 -- first. If one is found, there is an invariant that none
+                 -- of the instruction candidates without a ModR/M byte
+                 -- should validate with the set of parsed prefixes.
+                 Right (pfx, def) <- findDefWithPrefixBytes prefixBytes defsWithoutModRM
+              -> assert (all (\(mbVex, df) ->
+                               isLeft $ validatePrefixBytes prefixBytes mbVex df)
+                             defsWithModRM) $
+                 disassembleWithoutModRM def pfx
+
+              |  otherwise
+              -> disassembleWithModRM prefixBytes defsWithModRM
+
+    -- Disassemble instruction candidates with a ModR/M byte.
+    disassembleWithModRM :: [Word8]
+                         -> [(Maybe VEX, Def)]
+                         -> m InstructionInstance
+    disassembleWithModRM prefixBytes defs = do
+      tbl <- checkRequiredReg defs
+      case tbl of
+        RegTable v -> do
+          modRM <- readModRM
+          let mtbl = v V.! fromIntegral (modRM_reg modRM)
+          parseReadTable prefixBytes modRM (getReadTable modRM (getRMTable modRM mtbl))
+        RegUnchecked mtbl -> do
+          modRM <- readModRM
```

Comment:
Would lifting these `readModRM` calls out of the case be a bit clearer?

Reply:
Sure, why not. Done.